### PR TITLE
[Performance] 2D load lowering for column major B

### PR DESCRIPTION
**Pull request by chengjunlu**
Enable the Intel 2D block IO to load the B matrix resident on column major memory.

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 1]
```diff
@@ -366,7 +366,27 @@ struct LoadOpConversion
...
+    if (memoryLayoutInfor.getValue() == "column_major") {
+      memoryRowMajor = false;
+    }
+
     DotOperandEncodingAttr dotLayout = getDotEncoding(tensorType).value();
+    auto dotOrder = dotLayout.getThreadOrder();
+    bool valueRowMajor = true;
+    if (dotOrder[0] == 0 && dotOrder[1] == 1) {
+      valueRowMajor = false;
+    }
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:387]: should assert that `dotOrder[0] == 1 && dotOrder[1] == 0)`  when the condition at line 385 is false.

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 2]
```diff
@@ -445,48 +466,65 @@ struct LoadOpConversion
...
-      // PVC 2D load supports 64 bytes per row at most.
-      unsigned totalBytesPerRowPerDPASOp =
-          elemsPerDPASInst[1] * eltTy.getIntOrFloatBitWidth() / 8;
-      // Use block array length to load multiple operands B in N dim if
-      // possible.
-      vBlocks = std::min(repCluster[1], 64 / totalBytesPerRowPerDPASOp);
-      numOperandsOuterDimPerLoad = vBlocks;
+      if (isOperandA) {
+        return op.emitOpError("Transposing load doesn't support dot A layout.");
+      } else {
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:482]: Remove the else after a return.

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 3]
```diff
@@ -445,48 +466,65 @@ struct LoadOpConversion
...
+        if (!usePackedType)
+          return op.emitOpError(
+              "Transposing load doesn't support un-pack-able dot B layout.");
+
+        std::swap(tileHeight, tileWidth);
+
+        // we can decompose the matrix returned by transposed large 2d load
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:489]: [nit]: we -> We

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 4]
```diff
@@ -517,61 +564,84 @@ struct LoadOpConversion
           offsetY = add(offsetY, offsetBaseY);
           baseWidth = trunc(i32_ty, baseWidth);
           baseHeight = trunc(i32_ty, baseHeight);
-          rowStride = trunc(i32_ty, rowStride);
+          Value pitch = trunc(i32_ty, rowStride);
+          Value elemSizeInBytes = i32_val(originalElemBits / 8);
+
+          if (!memoryRowMajor) {
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:570]: why not check `memoryColumnMajor` here ?

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 5]
```diff
@@ -366,7 +366,27 @@ struct LoadOpConversion
     if (!hasDotDpasEncoding(tensorType))
       return failure();
 
+    Attribute blockIOAttr =
+        op->getAttr(TritonIntelGPUDialect::getBlockIOAttrName());
+    if (!blockIOAttr) {
+      return failure();
+    }
+
+    // Only support rand 2 dot layout. Either row major or column major
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:375]: rank?

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp (hunk 4 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:570]: `memoryColumnMajor` doesn't exist.

File: test/TritonIntelGPU/load-to-llvm-2dload.mlir [hunk 6]
```diff
@@ -179,10 +179,77 @@ module attributes {"triton_gpu.num-warps" = 8 : i32, "triton_gpu.threads-per-war
...
     %C = arith.constant dense<0.000000e+00> : tensor<32x32xf32, #dpas>
     %D = tt.dot %A, %B, %C, inputPrecision = tf32 : tensor<32x32xf16, #dot0> * tensor<32x32xf16, #dot1> -> tensor<32x32xf32, #dpas>
     tt.return
   }
 }
+
+// -----
+
+// CHECK:   llvm.func spir_funccc @llvm.genx.GenISA.LSC2DBlockRead.v16i32
+#mma = #triton_intel_gpu.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [1, 1], repCluster = [1, 2], A = [8, 16], B = [16, 32], C = [8, 32]}>
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply test/TritonIntelGPU/load-to-llvm-2dload.mlir]: done

File: test/TritonIntelGPU/load-to-llvm-2dload.mlir [hunk 7]
```diff
@@ -179,10 +179,77 @@ module attributes {"triton_gpu.num-warps" = 8 : i32, "triton_gpu.threads-per-war
...
+#dot_b = #triton_gpu.dot_op<{opIdx = 1, parent = #mma, kWidth = 2}>
+module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 1 : i32, triton_gpu.target = "xpu", "triton_gpu.threads-per-warp" = 16 : i32} {
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply test/TritonIntelGPU/load-to-llvm-2dload.mlir]: done

File: test/TritonIntelGPU/load-to-llvm-2dload.mlir [hunk 8]
```diff
@@ -179,10 +179,77 @@ module attributes {"triton_gpu.num-warps" = 8 : i32, "triton_gpu.threads-per-war
...
+  // CHECK-LABEL:   llvm.func spir_kernelcc @column_major_dot_b
+  tt.func public @column_major_dot_b(%arg0: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %col_stride: i32 {tt.divisibility = 16 : i32}) {
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply test/TritonIntelGPU/load-to-llvm-2dload.mlir]: done

File: test/TritonIntelGPU/load-to-llvm-2dload.mlir [hunk 9]
```diff
@@ -179,10 +179,77 @@ module attributes {"triton_gpu.num-warps" = 8 : i32, "triton_gpu.threads-per-war
...
+      // CHECK:    %[[VAL_176:.*]] = llvm.shufflevector %[[VAL_173]], %[[VAL_173]] [1, 3, 5, 7, 9, 11, 13, 15] : vector<16xi32>
+      %45 = tt.load %21 {triton_intel_gpu.block_io = "column_major"} : !tt.ptr<tensor<64x32xf16, #triton_gpu.dot_op<{opIdx = 1, parent = #mma, kWidth = 2}>>>
+      tt.return
+  }
+}
+
+// -----
+
+// CHECK:   llvm.func spir_funccc @_Z51intel_sub_group_2d_block_read_transpose_32b_16r8x1cPU3AS1viiiDv2_iPj
+#mma = #triton_intel_gpu.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [1, 1], repCluster = [1, 1], A = [8, 16], B = [16, 16], C = [8, 16]}>
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply test/TritonIntelGPU/load-to-llvm-2dload.mlir]: done

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp]: done

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 10]
```diff
@@ -366,7 +366,27 @@ struct LoadOpConversion
...
+    auto memoryLayoutInfor = cast<StringAttr>(blockIOAttr);
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp]: done

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 11]
```diff
@@ -366,7 +366,27 @@ struct LoadOpConversion
...
+    bool memoryRowMajor = true;
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp]: done

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp (hunk 2 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp]: done

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp (hunk 3 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp]: done

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp]: done

File: test/TritonIntelGPU/load-to-llvm-2dload.mlir (hunk 8 above)
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply test/TritonIntelGPU/load-to-llvm-2dload.mlir]: removed.

**Review by whitneywhtsang — APPROVED** [review]

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 12]
```diff
@@ -366,7 +366,25 @@ struct LoadOpConversion
...
+        op->getAttr(TritonIntelGPUDialect::getBlockIOAttrName());
+    if (!blockIOAttr)
+      return failure();
+
+    // Only support rank 2 dot layout, either row major or column major.
+    StringRef memoryLayoutInfo = cast<StringAttr>(blockIOAttr).getValue();
+    assert((memoryLayoutInfo == "row_major" ||
+            memoryLayoutInfo == "column_major") &&
+           "Only row_major or column_major is supported");
+    bool memoryRowMajor = (memoryLayoutInfo == "row_major");
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:379]: `bool` -> `const bool`

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 13]
```diff
@@ -366,7 +366,25 @@ struct LoadOpConversion
...
+
     DotOperandEncodingAttr dotLayout = getDotEncoding(tensorType).value();
+    auto dotOrder = dotLayout.getThreadOrder();
+    bool valueRowMajor = (dotOrder[0] == 1 && dotOrder[1] == 0);
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:383]: `bool` -> `const bool`

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 14]
```diff
@@ -366,7 +366,25 @@ struct LoadOpConversion
...
+    assert((valueRowMajor || (dotOrder[0] == 0 && dotOrder[1] == 1)) &&
+           "Only row_major or column_major is allowed");
+    bool isTransposeRequired = valueRowMajor ^ memoryRowMajor;
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:386]: here too

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp (hunk 4 above)
Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:567]: ok added more comments.

**Review by etiotto — APPROVED** [review]

**Pull request merged**